lib/shelley/test/unit/Cardano/Wallet/Shelley/MinimumUTxOSpec.hs

```diff
@@ -0,0 +1,166 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE TypeApplications #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module Cardano.Wallet.Shelley.MinimumUTxOSpec
+    ( spec
+    ) where
+
+import Prelude
+
+import Cardano.Api.Gen
+    ( genAddressShelley )
+import Cardano.Wallet.Primitive.Types.Address
+    ( Address (..) )
+import Cardano.Wallet.Primitive.Types.Coin
+    ( Coin (..) )
+import Cardano.Wallet.Primitive.Types.Coin.Gen
+    ( chooseCoin, shrinkCoin )
+import Cardano.Wallet.Primitive.Types.MinimumUTxO
+    ( MinimumUTxO
+    , MinimumUTxOForShelleyBasedEra (..)
+    , minimumUTxOForShelleyBasedEra
+    )
+import Cardano.Wallet.Primitive.Types.MinimumUTxO.Gen
+    ( genMinimumUTxO
+    , genMinimumUTxOForShelleyBasedEra
+    , shrinkMinimumUTxO
+    , shrinkMinimumUTxOForShelleyBasedEra
+    )
+import Cardano.Wallet.Primitive.Types.TokenBundle
+    ( TokenBundle (..) )
+import Cardano.Wallet.Primitive.Types.TokenMap
+    ( TokenMap )
+import Cardano.Wallet.Primitive.Types.TokenMap.Gen
+    ( genTokenMap, shrinkTokenMap )
+import Cardano.Wallet.Primitive.Types.Tx
+    ( TxOut (..) )
+import Cardano.Wallet.Shelley.Compatibility
+    ( fromCardanoAddress, toCardanoTxOut )
+import Cardano.Wallet.Shelley.MinimumUTxO
+    ( computeMinimumCoinForUTxO
+    , unsafeLovelaceToWalletCoin
+    , unsafeValueToLovelace
+    )
+import Data.Function
+    ( (&) )
+import Data.IntCast
+    ( intCast )
+import Data.Word
+    ( Word64 )
+import Generics.SOP
+    ( NP (..) )
+import Numeric.Natural
+    ( Natural )
+import Test.Hspec
+    ( Spec, describe, it )
+import Test.QuickCheck
+    ( Arbitrary (..), Property, property )
+import Test.QuickCheck.Classes
+    ( eqLaws, showLaws )
+import Test.QuickCheck.Extra
+    ( genericRoundRobinShrink, report, (<:>), (<@>) )
+import Test.Utils.Laws
+    ( testLawsMany )
+
+import qualified Cardano.Api.Shelley as Cardano
+import qualified Cardano.Wallet.Primitive.Types.TokenBundle as TokenBundle
+import qualified Data.ByteString as BS
+
+spec :: Spec
+spec = do
+    describe "Class instances obey laws" $ do
+        testLawsMany @MinimumUTxO
+            [ eqLaws
+            , showLaws
+            ]
+
+    describe "computeMinimumCoinForUTxO" $ do
+        it "prop_computeMinimumCoinForUTxO" $
+            prop_computeMinimumCoinForUTxO
+                & property
+        it "prop_computeMinimumCoinForUTxO_shelleyBasedEra_lowerBound" $
+            prop_computeMinimumCoinForUTxO_shelleyBasedEra_lowerBound
+                & property
+
+prop_computeMinimumCoinForUTxO :: MinimumUTxO -> TokenMap -> Property
+prop_computeMinimumCoinForUTxO minimumUTxO m = property $
+    computeMinimumCoinForUTxO minimumUTxO m >= Coin 0
+
+prop_computeMinimumCoinForUTxO_shelleyBasedEra_lowerBound
+    :: TokenBundle
+    -> Cardano.Address Cardano.ShelleyAddr
```

Comment:
What about byron addresses? (I'm slightly worried about these)

Reply:
> What about byron addresses? (I'm slightly worried, or at least certainly curious about these)

I'll investigate, and report back.

Reply:
> What about byron addresses? (I'm slightly worried, or at least certainly curious about these)

I've revised the property to also test with Byron addresses. See 91217999a78e6b9ff2cd3ef1206435865587d4b9.

It currently still passes.